lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -1549,45 +1549,53 @@ void PatchEntryPointMutate::addSpecialUserDataArgs(SmallVectorImpl<UserDataArg>
-          specialUserDataArgs.push_back(UserDataArg(builder.getInt32Ty(), "streamOutControlBuf",
-                                                    UserDataMapping::StreamOutControlBuf,
-                                                    &intfData->entryArgIdxs.tes.streamOutData.controlBufPtr));
-        }
         break;
       case ShaderStageGeometry:
-        if (m_pipelineState->getTargetInfo().getGfxIpVersion().major <= 10) {
-          // Allocate dummy stream-out register for geometry shader
-          userDataArgs.push_back(UserDataArg(builder.getInt32Ty(), "dummyStreamOut"));
-        } else if (m_pipelineState->enableSwXfb()) {
+        if (m_pipelineState->enableSwXfb()) {
           userDataArgs.push_back(UserDataArg(builder.getInt32Ty(), "streamOutTable", UserDataMapping::StreamOutTable,
                                              &intfData->entryArgIdxs.gs.streamOutData.tablePtr));
-          // NOTE: For GFX11+, the SW stream-out needs an additional special user data SGPR to store the
-          // stream-out control buffer address.
-          specialUserDataArgs.push_back(UserDataArg(builder.getInt32Ty(), "streamOutControlBuf",
-                                                    UserDataMapping::StreamOutControlBuf,
-                                                    &intfData->entryArgIdxs.gs.streamOutData.controlBufPtr));
+        } else {
+          assert(m_pipelineState->getTargetInfo().getGfxIpVersion().major <= 10);
+          // Allocate dummy stream-out register for geometry shader
+          userDataArgs.push_back(UserDataArg(builder.getInt32Ty(), "dummyStreamOut"));
         }
+          
         break;
       default:
         llvm_unreachable("Should never be called!");
         break;
       }
     }
   }
+
+  // NOTE: For GFX11+, the SW stream-out needs an additional special user data SGPR to store the stream-out control
+  // buffer address.
+  if (m_pipelineState->enableSwXfb()) {
+    switch (m_shaderStage) {
+    case ShaderStageVertex:
+      specialUserDataArgs.push_back(UserDataArg(builder.getInt32Ty(), "streamOutControlBuf",
+                                                UserDataMapping::StreamOutControlBuf,
+                                                &intfData->entryArgIdxs.vs.streamOutData.controlBufPtr));
+      break;
+    case ShaderStageTessEval:
+      specialUserDataArgs.push_back(UserDataArg(builder.getInt32Ty(), "streamOutControlBuf",
+                                                UserDataMapping::StreamOutControlBuf,
+                                                &intfData->entryArgIdxs.tes.streamOutData.controlBufPtr));
+      break;
+    case ShaderStageGeometry:
+      specialUserDataArgs.push_back(UserDataArg(builder.getInt32Ty(), "streamOutControlBuf",
+                                                UserDataMapping::StreamOutControlBuf,
+                                                &intfData->entryArgIdxs.gs.streamOutData.controlBufPtr));
+      break;
```

Comment:
Can this be factored further? The only difference here is which field of entryArgIdxs is written, everything else is the same.

Reply:
Sure. Will do.